File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/CTGeofenceReceiverTest.java
package com.clevertap.android.geofence;

import static org.awaitility.Awaitility.await;
import static org.mockito.Mockito.*;

import android.content.BroadcastReceiver;
import android.content.Intent;
import org.junit.*;
import org.mockito.*;

public class CTGeofenceReceiverTest extends BaseTestCase {

    @Mock
    public CTGeofenceAPI ctGeofenceAPI;

    @Mock
    public BroadcastReceiver.PendingResult pendingResult;

    @Mock
    private Logger logger;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.openMocks(this);
        super.setUp();
    }

    @Test
    public void testOnReceiveWhenIntentIsNull() {
        CTGeofenceReceiver receiver = new CTGeofenceReceiver();
        CTGeofenceReceiver spy = Mockito.spy(receiver);
        spy.onReceive(application, null);
        verify(spy, never()).goAsync();
    }

    @Test
    public void testOnReceiveWhenIntentNotNull() {
        CTGeofenceReceiver receiver = new CTGeofenceReceiver();
        CTGeofenceReceiver spy = Mockito.spy(receiver);
        when(spy.goAsync()).thenReturn(pendingResult);

        final Boolean[] isFinished = {false};

        doAnswer(invocation -> {
            isFinished[0] = true;
            return null;
        }).when(pendingResult).finish();

        try (MockedStatic<CTGeofenceAPI> ctGeofenceAPIMockedStatic = Mockito.mockStatic(CTGeofenceAPI.class)) {
            ctGeofenceAPIMockedStatic.when(CTGeofenceAPI::getLogger).thenReturn(logger);

            Intent intent = new Intent();
            spy.onReceive(application, intent);

            await().until(() -> isFinished[0]);

            verify(CTGeofenceAPI.getLogger()).debug(CTGeofenceAPI.GEOFENCE_LOG_TAG, "Geofence receiver called");
            verify(CTGeofenceAPI.getLogger()).debug(CTGeofenceAPI.GEOFENCE_LOG_TAG, "Returning from Geofence receiver");
            verify(pendingResult).finish();
        }
    }

}

File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/CTGeofenceReceiverTest.kt
package com.clevertap.android.geofence

import android.content.BroadcastReceiver
import android.content.Intent
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import io.mockk.spyk
import io.mockk.verify
import io.mockk.verifyOrder
import org.awaitility.Awaitility.await
import org.junit.Before
import org.junit.Test

class CTGeofenceReceiverTest : BaseTestCase() {

    private lateinit var ctGeofenceAPI: CTGeofenceAPI
    private lateinit var pendingResult: BroadcastReceiver.PendingResult
    private lateinit var logger: Logger

    @Before
    override fun setUp() {
        super.setUp()
        ctGeofenceAPI = mockk(relaxed = true)
        pendingResult = mockk(relaxed = true)
        logger = mockk(relaxed = true)
    }

    @Test
    fun testOnReceiveWhenIntentIsNull() {
        val receiver = CTGeofenceReceiver()
        val spy = spyk(receiver)

        spy.onReceive(application, null)

        verify(exactly = 0) { spy.goAsync() }
    }

    @Test
    fun testOnReceiveWhenIntentNotNull() {
        val receiver = CTGeofenceReceiver()
        val spy = spyk(receiver)
        every { spy.goAsync() } returns pendingResult

        val isFinished = arrayOf(false)

        every { pendingResult.finish() } answers {
            isFinished[0] = true
        }

        mockkStatic(CTGeofenceAPI::class) {
            every { CTGeofenceAPI.getLogger() } returns logger

            val intent = Intent()
            spy.onReceive(application, intent)

            await().until { isFinished[0] }

            verifyOrder {
                logger.debug(
                    CTGeofenceAPI.GEOFENCE_LOG_TAG, "Geofence receiver called"
                )
                logger.debug(
                    CTGeofenceAPI.GEOFENCE_LOG_TAG, "Returning from Geofence receiver"
                )
            }
            verify { pendingResult.finish() }
        }
    }
}
